presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,137 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SystemSessionProperties.isEvaluateProjectOnValues;
+import static com.facebook.presto.spi.relation.ExpressionOptimizer.Level.EVALUATED;
+import static com.facebook.presto.spi.relation.ExpressionOptimizer.Level.OPTIMIZED;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.values;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * This optimizer looks for ProjectNode followed by a ValuesNode and get the ProjectNode Evaluated.
+ * <p/>
+ * Plan before optimizer:
+ * <pre>
+ * ProjectNode (outputVariables)
+ *   - ValuesNode
+ * </pre>
+ * <p/>
+ * Plan after optimizer:
+ * <pre>
+ * ValuesNode (outputVariables)
+ * </pre>
+ */
+public class EvaluateProjectOnValues
+        implements Rule<ProjectNode>
+{
+    private static final Pattern<ProjectNode> PATTERN = project().with(source().matching(values()));
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public EvaluateProjectOnValues(FunctionAndTypeManager functionAndTypeManager)
+    {
+        requireNonNull(functionAndTypeManager, "functionManager is null");
+        this.functionAndTypeManager = functionAndTypeManager;
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isEvaluateProjectOnValues(session);
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        Optional<PlanNode> optionalSource = context.getLookup().resolveGroup(projectNode.getSource()).findFirst();
+        if (!optionalSource.isPresent()) {
+            return Result.empty();
+        }
+        ValuesNode source = (ValuesNode) optionalSource.get();
+        Collection<RowExpression> projectRowExpressions = projectNode.getAssignments().getExpressions();
+        List<List<RowExpression>> rows = source.getRows();
+        List<VariableReferenceExpression> valuesOutputVariables = source.getOutputVariables();
+
+        //evaluate ProjectNode
+        List<List<Object>> rowValues = rows.stream().map(row -> {
+            //Prepare to set up variable resolver
+            Map<String, Object> valuesMapForResolver = new HashMap<>();
+            IntStream.range(0, row.size()).forEach(index -> {
+                RowExpression element = row.get(index);
+                RowExpressionInterpreter elementInterpreter = new RowExpressionInterpreter(element,
+                        functionAndTypeManager,
+                        context.getSession().toConnectorSession(),
+                        EVALUATED);
```

Comment:
Does this handle non-deterministic functions? I think we should not apply this rule for non-deterministic like 

select random(x) from (values 1,2) AS T(x)

Reply:
What's the reasoning not evaluate a non-deterministic function? Is there a big difference evaluating it on a worker later on vs the coordinator?

Reply:
In general there could be surprises so we generally short-circuit lot of optimizations when we see non-deterministic expressions. Also I'm not a fan of RowExpressionInterpreter :( it's implementing the "engine" for constants and I'm sure it works differently from the java and differently differently from native lol

Reply:
non-deterministic function excluded.

Reply:
please add a test like: select random(x) from (values 1,2) AS T(x)

Reply:
Fixed.